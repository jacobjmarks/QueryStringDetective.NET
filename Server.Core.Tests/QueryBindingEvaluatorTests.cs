namespace Server.Core.Tests;

public class QueryBindingEvaluatorTests
{
    public static readonly TheoryData<string> TestCases = new()
    {
        "c",
        "foo",
        "a&q=b&q=c",
        "null",
        "0",
        "1",
        "-1",
        "1&q=2&q=3",
        "1&q=2&q=.3",
        "-1&q=0&q=1",
        $"{sbyte.MinValue}",
        $"{sbyte.MaxValue}",
        $"{byte.MaxValue}",
        $"{short.MinValue}",
        $"{short.MaxValue}",
        $"{ushort.MaxValue}",
        $"{int.MinValue}",
        $"{int.MaxValue}",
        $"{long.MinValue}",
        $"{long.MaxValue}",
        $"{ulong.MaxValue}",
        Uri.EscapeDataString($"{double.MinValue}"),
        Uri.EscapeDataString($"{double.MaxValue}"),
        $"{decimal.MinValue}",
        $"{decimal.MaxValue}",
        Uri.EscapeDataString($"{float.MinValue}"),
        Uri.EscapeDataString($"{float.MaxValue}"),
    };

    [Theory]
    [MemberData(nameof(TestCases))]
    public async Task AssertBinding(string queryString)
    {
        using var bindingEvaluator = new QueryBindingEvaluator();
        var results = await bindingEvaluator.EvaluateAsync("?q=" + queryString);

        var content = "Input: " + queryString
            + "\n\nBindings:\n" + string.Join("\n", results.Where(r => !r.IsErroneous)
                    .Select(r => $"{r.Type,-10} {r.Result}"))
            + "\n\nErroneous:\n" + string.Join("\n", results.Where(r => r.IsErroneous)
                    .Select(r => $"{r.Type,-10} {r.Error!.Message}"));

        content.Should().MatchSnapshot(SnapshotNameExtension.Create(ToValidFileName(queryString)));
    }

    private static string ToValidFileName(string value)
    {
        foreach (var invalidChar in Path.GetInvalidFileNameChars())
            value = value.Replace(invalidChar, '_');

        return value;
    }
}