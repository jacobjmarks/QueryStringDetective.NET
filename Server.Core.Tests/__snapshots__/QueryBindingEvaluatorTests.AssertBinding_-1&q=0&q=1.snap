Input: -1&q=0&q=1

Bindings:

string
  MinimalApis
    r = "-1,0,1"
  Controllers
    r = "-1"

string?
  MinimalApis
    r = "-1,0,1"
  Controllers
    r = "-1"

string[]
  MinimalApis
    r = ["-1","0","1"]
  Controllers
    r = ["-1","0","1"]

char
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "char q" from "-1,0,1".)
  Controllers
    e = 400 Bad Request (The value '-1,0,1' is not valid.)

char?
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "Nullable<char> q" from "-1,0,1".)
  Controllers
    e = 400 Bad Request (The value '-1,0,1' is not valid.)

char[]
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "char[] q" from "-1".)
  Controllers
    e = 400 Bad Request (The value '-1' is not valid.)

int
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "int q" from "-1,0,1".)
  Controllers
    r = -1

int?
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "Nullable<int> q" from "-1,0,1".)
  Controllers
    r = -1

int[]
  MinimalApis
    r = [-1,0,1]
  Controllers
    r = [-1,0,1]

double
  MinimalApis
    r = -101
  Controllers
    r = -1

double?
  MinimalApis
    r = -101
  Controllers
    r = -1

double[]
  MinimalApis
    r = [-1,0,1]
  Controllers
    r = [-1,0,1]

float
  MinimalApis
    r = -101
  Controllers
    r = -1

float?
  MinimalApis
    r = -101
  Controllers
    r = -1

float[]
  MinimalApis
    r = [-1,0,1]
  Controllers
    r = [-1,0,1]

decimal
  MinimalApis
    r = -101
  Controllers
    r = -1

decimal?
  MinimalApis
    r = -101
  Controllers
    r = -1

decimal[]
  MinimalApis
    r = [-1,0,1]
  Controllers
    r = [-1,0,1]

bool
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "bool q" from "-1,0,1".)
  Controllers
    e = 400 Bad Request (The value '-1,0,1' is not valid.)

bool?
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "Nullable<bool> q" from "-1,0,1".)
  Controllers
    e = 400 Bad Request (The value '-1,0,1' is not valid.)

bool[]
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "bool[] q" from "-1".)
  Controllers
    e = 400 Bad Request (The value '-1' is not valid.)

uint
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "uint q" from "-1,0,1".)
  Controllers
    e = 400 Bad Request (The value '-1,0,1' is not valid.)

uint?
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "Nullable<uint> q" from "-1,0,1".)
  Controllers
    e = 400 Bad Request (The value '-1,0,1' is not valid.)

uint[]
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "uint[] q" from "-1".)
  Controllers
    e = 400 Bad Request (The value '-1' is not valid.)

long
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "long q" from "-1,0,1".)
  Controllers
    r = -1

long?
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "Nullable<long> q" from "-1,0,1".)
  Controllers
    r = -1

long[]
  MinimalApis
    r = [-1,0,1]
  Controllers
    r = [-1,0,1]

ulong
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "ulong q" from "-1,0,1".)
  Controllers
    e = 400 Bad Request (The value '-1,0,1' is not valid.)

ulong?
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "Nullable<ulong> q" from "-1,0,1".)
  Controllers
    e = 400 Bad Request (The value '-1,0,1' is not valid.)

ulong[]
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "ulong[] q" from "-1".)
  Controllers
    e = 400 Bad Request (The value '-1' is not valid.)

sbyte
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "sbyte q" from "-1,0,1".)
  Controllers
    r = -1

sbyte?
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "Nullable<sbyte> q" from "-1,0,1".)
  Controllers
    r = -1

sbyte[]
  MinimalApis
    r = [-1,0,1]
  Controllers
    r = [-1,0,1]

short
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "short q" from "-1,0,1".)
  Controllers
    r = -1

short?
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "Nullable<short> q" from "-1,0,1".)
  Controllers
    r = -1

short[]
  MinimalApis
    r = [-1,0,1]
  Controllers
    r = [-1,0,1]

ushort
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "ushort q" from "-1,0,1".)
  Controllers
    e = 400 Bad Request (The value '-1,0,1' is not valid.)

ushort?
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "Nullable<ushort> q" from "-1,0,1".)
  Controllers
    e = 400 Bad Request (The value '-1,0,1' is not valid.)

ushort[]
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "ushort[] q" from "-1".)
  Controllers
    e = 400 Bad Request (The value '-1' is not valid.)

byte
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "byte q" from "-1,0,1".)
  Controllers
    e = 400 Bad Request (The value '-1,0,1' is not valid.)

byte?
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "Nullable<byte> q" from "-1,0,1".)
  Controllers
    e = 400 Bad Request (The value '-1,0,1' is not valid.)

byte[]
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "byte[] q" from "-1".)
  Controllers
    e = 400 Bad Request (The value '-1,0,1' is not valid.)
